### ISSUE TITLE
WebGLRenderTargets shouldn't be treated like special Textures

### ISSUE THREAD

**Issue by vanruesc** [mrdoob/three.js#8615]
##### Preface

Following up on #8417.

This issue's purpose is to keep track of a problem that was identified during an attempt to refactor the `WebGLRenderer` in order to move away from treating render targets like special textures.

##### The goal

WebGLRenderTargets should not be used like Textures. Their `texture` property should be used instead.

```js
material.map = renderTarget.texture; // not material.map = renderTarget;
```

##### The first attempt

WebGLRenderer: e255f01f 
WebGLPrograms: ae22bc1f
WebGLRenderTargetCube: dcfcd4dd

##### Description of the problem 

By dropping the special treatment for WebGLRenderTargets, this [flipEnvMap check](https://github.com/mrdoob/three.js/commit/e255f01fef932d02abf130c85ddd788ee997b19f#diff-c0e88b98497597a015ecf238e91ac3a0L2019) and these cube render target checks: [1](https://github.com/mrdoob/three.js/pull/8417/commits/e255f01fef932d02abf130c85ddd788ee997b19f#diff-c0e88b98497597a015ecf238e91ac3a0L2518) | [2](https://github.com/mrdoob/three.js/pull/8417/commits/e255f01fef932d02abf130c85ddd788ee997b19f#diff-c0e88b98497597a015ecf238e91ac3a0L2564) become problematic.  

When texture properties of materials are always `Texture` instances, it's no longer possible to rely on `instanceof WebGLRenderTargetCube`. The only way to know whether the texture at hand belongs to a `WebGLRenderTargetCube ` is an [ugly flag](https://github.com/mrdoob/three.js/commit/dcfcd4dd0ce6d172edbf0c80db171a14916b7872#diff-5d5b0d48c8d18c359a7d79a655ff731eR12). Furthermore, the `WebGLRenderTargetCube` has a `Texture` instance because it extends `WebGLRenderTarget`. `CubeTexture` is therefore not involved in these checks, so `instanceof THREE.CubeTexture` is not an option.  

Shader texture uniforms are another problem, because they are updated every frame.

Providing the user with [warnings](https://github.com/mrdoob/three.js/pull/8417/commits/e255f01fef932d02abf130c85ddd788ee997b19f#diff-c0e88b98497597a015ecf238e91ac3a0R2519) after checking for `instanceof THREE.WebGLRenderTarget` shouldn't be done every frame, but I couldn't find a better place for this.

Any suggestions would be much appreciated. Don't hesitate to make your own PR if you find a solid solution to the problem. 

##### Three.js version

- [x] Dev
- [ ] r75

**Comment by tschw** [comment]
Just moved that code around in #8606 . Now we already know whether a texture is 2D or Cube (because WebGL gives us that info querying the uniforms of the program). The entry points (from `WebGLUniforms`) into the renderer are now `setTexture2D` (formerly `setTexture`) and `setTextureCube`.

> Furthermore, the WebGLRenderTargetCube has a Texture instance because it extends WebGLRenderTarget

A quite elegant solution (possible in JavaScript, but one that would break in C++) is to defer the creation of the value that is used for `.texture` to a method (e.g. `createTexture`). The constructor of `WebGLRenderTarget` would call that method and `WebGLRenderTargetCube` would override it to get a proper `CubeTexture`. 

I think, mipmapping should be switchable, and off by default for render targets, and the renderer-internal function `setCubeTexture` could really use a bit of a cleanup. 

> Shader texture uniforms are another problem, because they are updated every frame.

Could force `.needsUpdate` to `true` on the render target texture by setting a property.

> Providing the user with warnings after checking for instanceof THREE.WebGLRenderTarget shouldn't be done every frame, but I couldn't find a better place for this.

See [here](https://github.com/mrdoob/three.js/blob/dev/src/renderers/WebGLRenderer.js#L3027-L3040).

HTH to pull your PR up (bet I broke it).

**Comment by vanruesc** [comment]
Thanks for the heads-up! These changes might really help, will look into it as soon as I can.

**Comment by verteAzur** [comment]
What does it means from a user perspective when one sees this warning?

**Comment by mrdoob** [comment]
@verteAzur 

Instead of doing this...

```js
var renderTexture = new THREE.RenderTexture( 256, 256 );
var material = new THREE.MeshBasicMaterial( {
    map: renderTexture
} );
```

You should do this:

```js
var renderTexture = new THREE.RenderTexture( 256, 256 );
var material = new THREE.MeshBasicMaterial( {
    map: renderTexture.texture
} );
```

What code is triggering this?

**Comment by verteAzur** [comment]
Thank you for the info. It the only warning/error I have seen from r77 to r78 migration.

**Comment by WestLangley** [comment]
@mrdoob You mean `renderTarget` and `THREE.WebGLRenderTarget`.

**Comment by verteAzur** [comment]
@mrdoob 
> What code is triggering this?
I have to pin point the exact spot. It is in the last part of the large loading process though. I have updated the mirror and water shaders, thinking it could be coming from that piece of code, but in vain. I'll get back to you when I find the line...

**Comment by mrdoob** [comment]
> @mrdoob You mean `renderTarget` and `THREE.WebGLRenderTarget`.

Ops! Thanks! It was early in the morning... 😇

**Comment by verteAzur** [comment]
The code triggering the warning is vast. It comes from the dynamic terrain example prior v75. It could be in the long list of integrated shaders: maybe the blooming or the blurring one. Do you have any idea?

**Comment by mrdoob** [comment]
No... Shouldn't you be able to check the callstack?

**Comment by verteAzur** [comment]
re mrdoob: I have found the line that needed the "renderTarget.texture" update. It was located in the render member function of the THREE.ShaderPass.prototype. I did make the change, but the warning was still triggering. Then, I've decided to update all postprocessing classes and shaders (Convolution, Copy,Effectcomposer, Shaderpass, Maskpass, RenderPass, Bloompass, FilmPass) and it did work like a charm. Everything is working without any warning nor error. 

We did notice a few things while going through the update process that might help others: #9267

Thank you all for all!
